src/saveload/saveload.cpp

```diff
@@ -1217,141 +1218,144 @@ static void *IntToReference(size_t index, SLRefType rt)
-	if (_sl.need_length != NL_NONE) {
-		SlSetLength(SlCalcListLen(list));
-		/* Determine length only? */
-		if (_sl.need_length == NL_CALCLENGTH) return;
-	}
-
-	typedef std::list<void *> PtrList;
-	PtrList *l = (PtrList *)list;
-
 	switch (_sl.action) {
-		case SLA_SAVE: {
-			SlWriteUint32((uint32)l->size());
-
-			PtrList::iterator iter;
-			for (iter = l->begin(); iter != l->end(); ++iter) {
-				void *ptr = *iter;
-				SlWriteUint32((uint32)ReferenceToInt(ptr, conv));
-			}
+		case SLA_SAVE:
+			SlWriteUint32((uint32)ReferenceToInt(*(void **)ptr, (SLRefType)conv));
 			break;
-		}
 		case SLA_LOAD_CHECK:
-		case SLA_LOAD: {
-			size_t length = IsSavegameVersionBefore(SLV_69) ? SlReadUint16() : SlReadUint32();
-
-			/* Load each reference and push to the end of the list */
-			for (size_t i = 0; i < length; i++) {
-				size_t data = IsSavegameVersionBefore(SLV_69) ? SlReadUint16() : SlReadUint32();
-				l->push_back((void *)data);
-			}
+		case SLA_LOAD:
+			*(size_t *)ptr = IsSavegameVersionBefore(SLV_69) ? SlReadUint16() : SlReadUint32();
 			break;
-		}
-		case SLA_PTRS: {
-			PtrList temp = *l;
-
-			l->clear();
-			PtrList::iterator iter;
-			for (iter = temp.begin(); iter != temp.end(); ++iter) {
-				void *ptr = IntToReference((size_t)*iter, conv);
-				l->push_back(ptr);
-			}
+		case SLA_PTRS:
+			*(void **)ptr = IntToReference(*(size_t *)ptr, (SLRefType)conv);
 			break;
-		}
 		case SLA_NULL:
-			l->clear();
+			*(void **)ptr = nullptr;
 			break;
 		default: NOT_REACHED();
 	}
 }
 
-
 /**
- * Template class to help with std::deque.
+ * Template class to help with list-like types.
  */
-template <typename T>
-class SlDequeHelper {
-	typedef std::deque<T> SlDequeT;
+template <template<typename, typename> typename Tstorage, typename Tvar, typename Tallocator = std::allocator<Tvar>>
+class SlStorageHelper {
+	typedef Tstorage<Tvar, Tallocator> SlStorageT;
 public:
 	/**
-	 * Internal templated helper to return the size in bytes of a std::deque.
-	 * @param deque The std::deque to find the size of
+	 * Internal templated helper to return the size in bytes of a list-like type.
+	 * @param storage The storage to find the size of
 	 * @param conv VarType type of variable that is used for calculating the size
+	 * @param cmd The SaveLoadType ware are saving/loading.
 	 */
-	static size_t SlCalcDequeLen(const void *deque, VarType conv)
+	static size_t SlCalcLen(const void *storage, VarType conv, SaveLoadType cmd = SL_VAR)
 	{
-		const SlDequeT *l = (const SlDequeT *)deque;
+		assert(cmd == SL_VAR || cmd == SL_REF);
 
-		int type_size = 4;
-		/* Each entry is saved as type_size bytes, plus type_size bytes are used for the length
-		 * of the list */
-		return l->size() * SlCalcConvFileLen(conv) + type_size;
+		const SlStorageT *list = static_cast<const SlStorageT *>(storage);
+
+		int type_size = 4; // Size of the length of the list.
+		int item_size = cmd == SL_VAR ? SlCalcConvFileLen(conv) : (IsSavegameVersionBefore(SLV_69) ? 2 : 4);
```

Comment:
This seems to be used only for saving (`	assert(_sl.action == SLA_SAVE);` does not fail), so the version check is not needed.